src/capymoa/anomaly/online_isolation_forest.py

```diff
@@ -0,0 +1,341 @@
+
+    def recursive_unbuild(self, node: OnlineIsolationNode) -> OnlineIsolationNode:
+        # If the current node is a leaf, return it
+        if node.children is None:
+            return node
+        # If the current node is not a leaf, unbuild it
+        else:
+            # Recursively unbuild children
+            for i, _ in enumerate(node.children):
+                node.children[i]: OnlineIsolationNode = self.recursive_unbuild(node.children[i])
+            # Update the vectors of minimum and maximum values seen so far by the current node
+            node.min_values: ndarray = vstack([node.children[i].min_values for i, _ in enumerate(node.children)]).min(axis=0)
+            node.max_values: ndarray = vstack([node.children[i].max_values for i, _ in enumerate(node.children)]).max(axis=0)
+            # Delete children nodes, projection vector and split values
+            node.children: ndarray[OnlineIsolationNode] = None
+            node.projection_vector: ndarray[float] = None
+            node.split_values: ndarray[float] = None
+            return node
+
+    def predict(self, data: ndarray) -> ndarray[float]:
+        # Compute depth of each sample
+        if self.root:
+            return self.recursive_depth_search(self.root, data, empty(shape=(data.shape[0],), dtype=float))
+        else:
+            return zeros(shape=(data.shape[0],), dtype=float)
+
+    def recursive_depth_search(self, node: OnlineIsolationNode, data: ndarray, depths: ndarray[float]) -> ndarray[float]:
+        # If the current node is a leaf, fill the depths vector with the current depth plus a normalization factor
+        if node.children is None or data.shape[0] == 0:
+            depths[:] = node.depth + OnlineIsolationTree.get_random_path_length(self.branching_factor, self.max_leaf_samples,
+                                                                                node.data_size)
+        else:
+            # Partition data
+            partition_indices: list[ndarray[int]] = self.split_data(data, node.projection_vector, node.split_values)
+            # Fill the vector of depths
+            for i, indices in enumerate(partition_indices):
+                depths[indices]: ndarray[float] = self.recursive_depth_search(node.children[i], data[indices],
+                                                                              depths[indices])
+        return depths
+
+    def split_data(self, data: ndarray, projection_vector: ndarray[float], split_values: ndarray[float]) -> list[ndarray[int]]:
+        # Project data using projection vector
+        projected_data: ndarray = data @ projection_vector
+        # Sort projected data and keep sort indices
+        sort_indices: ndarray = argsort(projected_data)
+        # Split data according to their membership
+        partition: list[ndarray[int]] = split(sort_indices, projected_data[sort_indices].searchsorted(split_values))
+        return partition
+
+
+class OnlineIsolationNode:
```

Comment:
Consider using a data-class here https://docs.python.org/3/library/dataclasses.html.

Reply:
Many of these types should be `Optional[]` since they can be `None`. The issue is that this makes following the logic harder because the types are wrong. e.g `if node.children is None:` should always be `False`